### PR TITLE
[rb][bidi] Added print command with tests

### PR DESCRIPTION
**Pull request by TamsilAmani**
### Description

Print command for BiDi Browsing Context module along with tests.

### Motivation and Context

https://w3c.github.io/webdriver-bidi/#command-browsingContext-print

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: rb/spec/integration/selenium/webdriver/bidi/browsing_context_spec.rb [hunk 1]
```diff
@@ -23,6 +23,8 @@ module Selenium
   module WebDriver
     class BiDi
       describe BrowsingContext, only: {browser: %i[chrome edge firefox]} do
+        let(:magic_number) { 'JVBER' }
```
Comment (p0deje) [inline rb/spec/integration/selenium/webdriver/bidi/browsing_context_spec.rb:26]: What is this magic number?

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb [hunk 2]
```diff
@@ -73,6 +73,35 @@ def get_tree(max_depth: nil)
           )
         end
 
+        def print_page(**options)
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/bidi/browsing_context.rb]: I suggest we explicitly list out all available options in the method signature and then just refer to them

```ruby
def print_page(
  bottom: 1.0,
  top: 1.0,
  left: 1.0,
  right: 1.0,
  height: 27.94, 
  width: 21.59,
  page_ranages: [],
  scale: 1.0,
  shrink_to_fit: true
)
  print_result = @bidi.send_cmd(
    'browsingContext.print',
    context: @id,
    background: false,
    margin: {bottom: bottom, top: top, left: left, right: right},
    orientation: 'portrait',
    page: {height: height, width: width},
    pageRanges: page_ranges,
    scale: scale,
    shrinkToFit: shrink_to_fit
  )

  print_result['data']
end
```

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb [hunk 3]
```diff
@@ -73,6 +73,35 @@ def get_tree(max_depth: nil)
...
+                                        background: false,
+                                        margin: {
+                                          bottom: options['bottom'] || 1.0,
+                                          left: options['left'] || 1.0,
+                                          right: options['right'] || 1.0,
+                                          top: options['top'] || 1.0
+                                        },
+                                        orientation: 'portrait',
+                                        page: {
+                                          height: options.dig(:page, :height) || 27.94,
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/bidi/browsing_context.rb]: Why so precise default values? Is it something from the spec?

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb [hunk 4]
```diff
@@ -73,6 +73,35 @@ def get_tree(max_depth: nil)
...
+                                        },
+                                        pageRanges: options['page_ranges'] || [],
+                                        scale: options['scale'] || 1.0,
+                                        shrinkToFit: options['shrink_to_fit'] || true)
+
+          print_result['data']
+        end
+
+        def save_print_page(path, **options)
+          File.open(path, 'wb') do |file|
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/bidi/browsing_context.rb]: You can simplify this:

```ruby
File.write(path, Base64.decode64(print_page(**options)), mode: 'wb')
```

File: rb/spec/integration/selenium/webdriver/bidi/browsing_context_spec.rb (hunk 1 above)
Comment (titusfortner) [reply rb/spec/integration/selenium/webdriver/bidi/browsing_context_spec.rb:26]: We use it in our current print tests. I thought you knew what it was. I must have copied it from Java. /shrug

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb (hunk 2 above)
Comment (titusfortner) [reply rb/lib/selenium/webdriver/bidi/browsing_context.rb]: Should we create a basic PORO/Struct for options so we don't have so many optional parameters in this method? We can also use it for WebDriver classic print method.

Comment (p0deje) [reply rb/lib/selenium/webdriver/bidi/browsing_context.rb]: That would be nice actually. Structs would be enough:

```ruby
Margin = Struct.new(:bottom, :top, :left, :right, keyword_init: true) do 
  def initialize(bottom: 1.0, top: 1.0, left: 1.0, right: 1.0)
    super
  end
end

def print_page(margin: Margin.new, ...)
```

File: rb/spec/integration/selenium/webdriver/bidi/browsing_context_spec.rb (hunk 1 above)
Comment (TamsilAmani) [reply rb/spec/integration/selenium/webdriver/bidi/browsing_context_spec.rb:26]: _Magic number_ is the static base64 encoded bytes which appear at the start of files. For PDF it's value is 'JVBER'. That's why we use it to test here.
More info: 
https://bitofhex.com/2019/02/19/base64-encoded-signatures/

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb (hunk 3 above)
Comment (TamsilAmani) [reply rb/lib/selenium/webdriver/bidi/browsing_context.rb]: Yes, the default values are mentioned in the spec.
https://w3c.github.io/webdriver-bidi/#command-browsingContext-print

**Comment by TamsilAmani** [comment]
@titusfortner should I combine `save_print_page` with `print_page`? Since the former only needs one extra `path` parameter.

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb (hunk 2 above)
Comment (TamsilAmani) [reply rb/lib/selenium/webdriver/bidi/browsing_context.rb]: Done

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb (hunk 4 above)
Comment (TamsilAmani) [reply rb/lib/selenium/webdriver/bidi/browsing_context.rb]: Done

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb [hunk 5]
```diff
@@ -73,6 +73,37 @@ def get_tree(max_depth: nil)
...
+          end
+        end
+
+        def print_page(background: false, margin: Margin.new, orientation: 'portrait', page: Page.new, page_ranges: [],
+                       scale: 1.0, shrink_to_fit: true, path: nil)
+          print_result = @bidi.send_cmd('browsingContext.print',
+                                        context: @id,
+                                        background: background,
+                                        margin: {bottom: margin.bottom, top: margin.top, left: margin.left,
+                                                 right: margin.right},
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/bidi/browsing_context.rb]:
```suggestion
                                        margin: margin.to_h,
```

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb [hunk 6]
```diff
@@ -73,6 +73,37 @@ def get_tree(max_depth: nil)
...
+                                        orientation: orientation,
+                                        page: {height: page.height, width: page.width},
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/bidi/browsing_context.rb]:
```suggestion
                                        page: page.to_h,
```

File: rb/spec/integration/selenium/webdriver/bidi/browsing_context_spec.rb (hunk 1 above)
Comment (p0deje) [reply rb/spec/integration/selenium/webdriver/bidi/browsing_context_spec.rb:26]: TIL

**Review by p0deje — APPROVED** [review]
Left a couple of suggestion, but looks great overall

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb (hunk 5 above)
Comment (TamsilAmani) [reply rb/lib/selenium/webdriver/bidi/browsing_context.rb]: Done

File: rb/lib/selenium/webdriver/bidi/browsing_context.rb (hunk 6 above)
Comment (TamsilAmani) [reply rb/lib/selenium/webdriver/bidi/browsing_context.rb]: Done

**Pull request closed without merging**